crates/circuit/src/circuit_data.rs

```diff
@@ -152,18 +165,245 @@ pub struct CircuitData {
+        num_qubits: u32,
+        instructions: I,
+        global_phase: Param,
+    ) -> PyResult<Self>
+    where
+        I: IntoIterator<Item = (StandardGate, SmallVec<[Param; 3]>, SmallVec<[u32; 2]>)>,
+    {
+        let instruction_iter = instructions.into_iter();
+        let mut res = CircuitData {
+            data: Vec::with_capacity(instruction_iter.size_hint().0),
+            intern_context: InternContext::new(),
+            qubits_native: Vec::with_capacity(num_qubits as usize),
+            clbits_native: Vec::new(),
+            qubit_indices_native: HashMap::with_capacity(num_qubits as usize),
+            clbit_indices_native: HashMap::new(),
+            qubits: PyList::empty_bound(py).unbind(),
+            clbits: PyList::empty_bound(py).unbind(),
+            param_table: ParamTable::new(),
+            global_phase,
+        };
+        if num_qubits > 0 {
+            let qubit_cls = QUBIT.get_bound(py);
+            for _i in 0..num_qubits {
+                let bit = qubit_cls.call0()?;
+                res.add_qubit(py, &bit, true)?;
+            }
+        }
+        for (operation, params, qargs) in instruction_iter {
+            let qubits = PyTuple::new_bound(
+                py,
+                qargs
+                    .iter()
+                    .map(|x| res.qubits_native[*x as usize].clone_ref(py))
+                    .collect::<Vec<PyObject>>(),
+            )
+            .unbind();
+            let empty: [u8; 0] = [];
+            let clbits = PyTuple::new_bound(py, empty);
+            let inst = res.pack_owned(
+                py,
+                &CircuitInstruction {
+                    operation: OperationType::Standard(operation),
+                    qubits,
+                    clbits: clbits.into(),
+                    params,
+                    extra_attrs: None,
+                    #[cfg(feature = "cache_pygates")]
+                    py_op: None,
+                },
+            )?;
+            res.data.push(inst);
+        }
+        Ok(res)
+    }
+
+    /// Add an instruction's entries to the parameter table
+    fn update_param_table(
+        &mut self,
+        py: Python,
+        inst_index: usize,
+        _params: Option<Vec<(usize, Vec<PyObject>)>>,
+    ) -> PyResult<bool> {
+        if let Some(params) = _params {
+            let mut new_param = false;
+            for (param_index, raw_param_objs) in &params {
+                let atomic_parameters: HashMap<u128, PyObject> = raw_param_objs
+                    .iter()
+                    .map(|x| {
+                        (
+                            x.getattr(py, intern!(py, "_uuid"))
+                                .expect("Not a parameter")
+                                .getattr(py, intern!(py, "int"))
+                                .expect("Not a uuid")
+                                .extract::<u128>(py)
+                                .unwrap(),
+                            x.clone_ref(py),
```

Comment:
Heh, I did it specifically to avoid the extra overhead of a second `isinstance` check. We actually already did the type check on `FromPyObject` as to get to this code path we're in the `Param::ParameterExpression` variant.

Even ignoring that though this is pretty deep in in the internals of Qiskit and if someone created an object that had a `_uuid` attr which returned something with an `int` attr that returned a `u128` extractable object that they'd have to try pretty hard to make something look like a parameter which wasn't.

Reply:
I'm fine ignoring this.